### DISCUSSION TITLE
Kyuubi-ha 1.7.1 uses Zookeeper 3.4 as a dependency

### DISCUSSION THREAD

**Discussion by HaoYang670** [apache/kyuubi#5069] [Q&A]
Hi,

I am writing my own zookeeper client based on Kyuubi's `ZookeeperDiscoveryClient` 1.7.1 and java 17.

It is now using zoopeeker 3.4 as a denpendency: https://mvnrepository.com/artifact/org.apache.kyuubi/kyuubi-ha_2.12/1.7.1, which leads me to get `Session 0x0 for server localhost/<unresolved>:2181, unexpected error` at runtime.

Is there a way to upgrade the version to 3.6.2 at runtime?

**Comment by pan3793** [comment]
Please check the https://github.com/apache/kyuubi/issues/4854

**Reply by HaoYang670** [reply]
Ok. I have switched to 3.6.3, but got this error:
```
Session 0x0 for sever localhost/127.0.0.1:2181, Closing socket connection. Attempting reconnect except it is a SessionExpiredException.

java.net.ConnectException: Connection refused
        at java.base/sun.nio.ch.Net.pollConnect(Native Method) ~[na:na]
        at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672) ~[na:na]
        at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:946) ~[na:na]
        at org.apache.zookeeper.ClientCnxnSocketNIO.doTransport(ClientCnxnSocketNIO.java:344) ~[zookeeper-3.6.3.jar:3.6.3]
        at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1290) ~[zookeeper-3.6.3.jar:3.6.3]
```
This is how I create client in my code
```java
        final KyuubiConf conf = new KyuubiConf(false);
        conf.set("kyuubi.ha.addresses", "localhost:2181");
        this.client = new ZookeeperDiscoveryClient(conf);
```

**Reply by pan3793** [reply]
I'm not sure what changes you have done for upgrading Zookeeper 3.6.3, but I suppose simply changing the `zookeeper.version` defined at `pom.xml` would not work.

**Reply by HaoYang670** [reply]
I just add zookeeper as a dependency
```
				<!-- https://mvnrepository.com/artifact/org.apache.kyuubi/kyuubi-ha -->
				<dependency>
					<groupId>org.apache.kyuubi</groupId>
					<artifactId>kyuubi-ha_2.12</artifactId>
					<version>1.7.1</version>
				</dependency>
				<!-- https://mvnrepository.com/artifact/org.apache.zookeeper/zookeeper -->
				<dependency>
					<groupId>org.apache.zookeeper</groupId>
					<artifactId>zookeeper</artifactId>
					<version>3.6.3</version>
				</dependency>
```

**Reply by pan3793** [reply]
That does not work ... have you maintained the internal fork? consider cherry-picking the master JDK-17 related commits to your internal branch. Or waiting for the 1.8.0 release(early Aug. hopeful)

**Reply by HaoYang670** [reply]
> that does not work ...

But why does the error message change?

**Reply by pan3793** [reply]
The ZK 3.4 does not work with JDK 17 because of ZOOKEEPER-3779, it was explained in https://github.com/apache/kyuubi/pull/4852. The ZK 3.6 failure likes a normal socket connect refuse error, I can not come to a conclusion with only such an error stack. Have you seen that the server starts success and listens to `localhost:2181`, can it be connected via `zkCli.sh`?

**Reply by HaoYang670** [reply]
> Have you seen that the server starts success and listens to localhost:2181

I think so, let me check it later. I ran a kyuubi server in K8s and I port-forward the 2181 port to localhost

**Reply by HaoYang670** [reply]
Thanks @pan3793. I solved the problem after downgrading to java 8. Thanks for your help.
